From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ian Wahbe <ian@wahbe.com>
Date: Thu, 11 May 2023 18:11:47 -0700
Subject: [PATCH] expose providers


diff --git a/shim/shim.go b/shim/shim.go
new file mode 100644
index 000000000..5f45ebd86
--- /dev/null
+++ b/shim/shim.go
@@ -0,0 +1,12 @@
+package shim
+
+import (
+	"github.com/hashicorp/terraform-plugin-framework/provider"
+
+	"github.com/cloudflare/terraform-provider-cloudflare/internal"
+)
+
+
+func PFProvider() provider.Provider {
+	return internal.NewProvider("pulumi")()
+}
