From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ian Wahbe <ian@wahbe.com>
Date: Thu, 11 May 2023 18:11:47 -0700
Subject: [PATCH] expose providers


diff --git a/shim/shim.go b/shim/shim.go
new file mode 100644
index 00000000..15d62e4b
--- /dev/null
+++ b/shim/shim.go
@@ -0,0 +1,17 @@
+package shim
+
+import (
+	"github.com/hashicorp/terraform-plugin-framework/provider"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	pf "github.com/cloudflare/terraform-provider-cloudflare/internal/framework/provider"
+	sdk "github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider"
+)
+
+func SDKProvider() *schema.Provider {
+	return sdk.New("")()
+}
+
+func PFProvider() provider.Provider {
+	return pf.New("")()
+}
